### PR TITLE
Datadog Logs Module

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Create the `DatadogLogs` module.

### How?

> I'm sorry for the size of this PR, it's partially due to the xcproj but also because I've migrated part of the unit test. Some code needed to move to `DatatogInternal` and `TestUtilities`.

#### `DatatogInternal` utils
The following definitions are now part of `DatadogInternal`
- `Sampler`
- `RelativeDateProvider`
- `DDError`
- `ProgrammerError`
- `InternalError`
- `SwiftExtensions.swift`

#### `DatadogLogger`
The `DatadogLogger` is now the entry point of the Logs Feature. 

##### Configuration

The public initialise methods creates and register the Logs Feature to the given core:
```swift
public static func initialise(
    in core: DatadogCoreProtocol = defaultDatadogCore,
    uploadURL: URL,
    applicationBundleIdentifier: String,
    eventMapper: LogEventMapper? = nil,
    dateProvider: DateProvider = SystemDateProvider(),
    sampler: Sampler
) throws
```
This signature is not yet user-friendly and it will be revisited when we tackle configuration. Currently, only the `Datadog` module uses it.

##### Builder
The `DatadogLogger` also exposes a `builder` static variable to configure and build a logger instance. This builder pattern stays untouch but will be revisited for a more idiomatic setup.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [x] Run smoke tests

### CODE REVIEW

File: DatadogLogs/Sources/Builder.swift
Comment (maxep) [inline DatadogLogs/Sources/Builder.swift:18]: This is exact same builder as in v1.

File: TestUtilities/Helpers/TestsDirectory.swift
Comment (maxep) [inline TestUtilities/Helpers/TestsDirectory.swift:13]: The temporary directory has been moved to `TestUtilities` but without the dependency to `Directory` or `File` from the v1 module. By the end of the migration I'm hoping that this logic will be used by the IntegrationTests only.

File: DatadogInternal/Sources/Utils/SwiftExtensions.swift
Comment (maxep) [inline DatadogInternal/Sources/Utils/SwiftExtensions.swift:99]: This is copied from SR 👍

File: Tests/DatadogTests/Matchers/RUMSessionMatcher.swift
Comment (maxep) [inline Tests/DatadogTests/Matchers/RUMSessionMatcher.swift:91]: Dates in models are milliesec in `Int64`

File: Sources/Datadog/Datadog.swift
Comment (maxep) [inline Sources/Datadog/Datadog.swift:108]: random values

Comment (maxep) [inline Sources/Datadog/Datadog.swift:108]: Note that verbosity now uses levels from core and no longer from the Log Feature.

**Review by ncreated — COMMENTED** [review]
It looks good 👍. Due to the size of change I couldn't go deep into how we solved individual problems, but I asked some clarifying questions. Waiting for replies before approval 🙂.

File: DatadogLogs/Sources/Feature/DatadogLogsFeature.swift
Comment (maxep) [reply DatadogLogs/Sources/Feature/DatadogLogsFeature.swift:10]: Yes, I'm also thinking of making the Feature name a static var in the `DatadogFeature` protocol, so then we would do `DatadogLogsFeature.name`, it would also simplify registration. In the meantime, I will commit your second suggestion 👍

File: DatadogLogs/Sources/Logger.swift
Comment (maxep) [reply DatadogLogs/Sources/Logger.swift:31]: Yes, this `Logger` won't be used as a type, only `DatadogLogger`.

File: DatadogLogs/Sources/Scrubbing/LogEventMapper.swift
Comment (maxep) [reply DatadogLogs/Sources/Scrubbing/LogEventMapper.swift:27]: Indeed, I will check!

File: DatadogSDK.podspec
Comment (maxep) [reply DatadogSDK.podspec:30]: This is not the core yet, We need that dependency for initialisation. The configuration part will be tackled separately.

File: DatadogLogs/Sources/Scrubbing/LogEventMapper.swift
Comment (maxep) [reply DatadogLogs/Sources/Scrubbing/LogEventMapper.swift:27]: So this class is used by the `Datadog.Configuration.Builder` in the `Datadog` module. So this class will be made internal when we split the config 👍

**Comment by maxep** [comment]
Thanks for the review @ncreated and sorry again for this size. I've answered your comments, let me know if you need more clarification 👍

File: DatadogLogs/Sources/Logger.swift
Comment (ncreated) [reply DatadogLogs/Sources/Logger.swift:31]: Even then, the problem will be still there:
```swift
import DatadogLogs

let logger: Logger
```
☝️ It could be still ambiguous because `Logger` could either mean `DatadogLogs.Logger` protocol or `Logger` struct/enum/protocol/class defined in app module. I think there is no other way around than changing the name if we want to keep it in public interface. `DDLogger` could be a simple fix there - we could tackle it in separate JIRA task. WDYT?

File: DatadogLogs/Sources/Scrubbing/LogEventMapper.swift
Comment (ncreated) [reply DatadogLogs/Sources/Scrubbing/LogEventMapper.swift:27]: Got it 👍, please track it in JIRA 👌.

File: DatadogSDK.podspec
Comment (ncreated) [reply DatadogSDK.podspec:30]: Oh got it - it's transient and only for migration.

**Review by ncreated — APPROVED** [review]
Looks good 👍, thanks for answers 🙂.

**Pull request merged**
